### PR TITLE
Faster init with LoRA

### PR DESCRIPTION
**Pull request by poedator**
Purpose: accelerate peft model initialisation with LoRA config.
see issue description in https://github.com/huggingface/peft/issues/871

The idea is to avoid costly calls of weight initialisation functions in CPU, when those weights are not necessary and will be replaced by the original weights anyways.

This PR initially changes only init for Linear layers. 

potential further improvements:
- apply to Embedding and Conv2d layers
- apply to compressed layer types, like Linear4bit, Linear8bit
- init LoRA modules (A, B) right on target device, not CPU

### CODE REVIEW

**Comment by poedator** [comment]
> I wonder if we can use [`skip_init`](https://pytorch.org/docs/stable/generated/torch.nn.utils.skip_init.html) or something else from PyTorch instead of monkey patching a hard-coded list of initialization methods.

well, I borrowed this idea [from GPTQ code](https://github.com/IST-DASLab/gptq/blob/2d65066eeb06a5c9ff5184d8cebdf33662c67faf/llama.py#L15). But I support your idea to avoid monkey patching.
Let me try this improvement - I will write back soon.

**Comment by poedator** [comment]
@BenjaminBossan !
`skip_init()` turned out not very convenient to apply from inside of a subclass. However, looking at its source code, I got an idea to use `meta` device. It works much faster than my original code, since meta tensors are faster than empty cpu tensors. Thanks for the suggestion!
To accomodate this change I also:
	- made `self.to(self.weight.device)` in LoraLayer.update_layer() conditional, to avoid moving loras to 'meta' device
	- removed  self.weight.requires_grad = False in `Linear.__init__()` for it is not relevant.   # lora.py:850

Now with Llama-2-7B application of Guanaco LoRA adapter takes 1.5 seconds vs 103 seconds originally.

**Comment by BenjaminBossan** [comment]
@poedator Great, thanks for experimenting. This looks like a very nice speedup indeed. I'm not 100% sure if we can remove all of that code, but let's see. Could you please run `make style` on your code? Then the code quality checks should pass and the unit tests should run.

**Comment by poedator** [comment]
I added couple more commits to pass device param to LoRA layer constructors to initialize LoRA adapters right at the target device. Apparently I need to do some more local testing ...
As a matter of introduction: I am a junior researcher @ Yandex, 2d name in this paper: https://arxiv.org/abs/2306.03078. We use HF packages daily and want them to get even better ).

**Comment by BenjaminBossan** [comment]
Thanks for applying the fixes and providing more context. 

Unfortunately, the tests are failing right now. The main reason seems to be that the code path when `init_lora_weights=False` is not properly implemented yet, resulting in the weights staying on meta device. This needs to be fixed and it might require a bit of fiddling, as the logic is a bit all over the place. It would probably help if we cleaned up the initialization logic, but that requires some non-trivial refactoring.

To proceed, would you be up for fixing the failing tests for the code path I mentioned?

**Comment by poedator** [comment]
@BenjaminBossan 
I updated the PR to catch up with #807 and also limited my changes to just the Linear layer init to prove the concept. 
Some of the tests with `init_lora_weights=False` are also failing locally (test_merge_layers_09 and 10). This is puzzling because they do not seem to involve LoRA for Linear. Nevertheless I want to see this PR coming thru, so let me try to fix it with your guidance.

pls be more specific on the failing tests. do you want me to update the tests code or to change some LoRA code?

**Comment by BenjaminBossan** [comment]
Thanks for the updates and the further investigation. Regarding the failing tests, the idea is that all existing tests should pass after the change. There could be tests that need adjustment, but I'm not aware of any. Anyway, for me, all the tests passed. As you can see, the CI passes too. So not sure why they failed for you.

I think the strategy you suggested, which is to start with one layer type first, is valid. We can do that and once that has proven to work well, we can tackle the others.

I did some further digging into the code and IIUC, I think the weight initialization happening in these two lines is completely unnecessary:

https://github.com/huggingface/peft/blob/0b2f950cc212dd8acda983760c0aba2bf4872a00/src/peft/tuners/lora/layer.py#L142
https://github.com/huggingface/peft/blob/0b2f950cc212dd8acda983760c0aba2bf4872a00/src/peft/tuners/lora/layer.py#L151

It creates a randomly initialized `self.weight` (and possibly `self.bias`) on the lora layer, which is then initialized with kaiming init twice! And later, it is replaced by the original weight of the target layer, so the whole initialization was unnecessary:

https://github.com/huggingface/peft/blob/0b2f950cc212dd8acda983760c0aba2bf4872a00/src/peft/tuners/lora/model.py#L213

Therefore, I think this step could be completely skipped. Since we're talking about the full sized weight here, not the smaller LoRA weights, this could take quite some time. The fix you proposed would create that weight on the meta device, but I wonder if we can rewrite the code to completely skip the initialization instead. If you want to explore that, feel free to do so, otherwise I'll take a look at it later. If you have some benchmark script to test the initialization speed, please share it.

> Also I may try to pass device argument to LoRAs inits so they are created on Cuda where they belong (still faster). But that may be better left to a separate PR.

Yes, that's indeed best left for a separate PR.

**Comment by BenjaminBossan** [comment]
@poedator I attempted to remove the redundant initializations completely in #887. If you have time to test potential speedups of that branch, it would be great.

**Comment by poedator** [comment]
> @poedator I attempted to remove the redundant initializations completely in #887. If you have time to test potential speedups of that branch, it would be great.

Thanks for the initiative. Let me test #887 later today.

**Comment by BenjaminBossan** [comment]
Thanks for providing the script @poedator. Could you please paste your measurements for reference? Unfortunately, I still haven't gotten access to Llama 2 (really not sure why it's still pending), so I can't test that model. For the others, I got:

```
# main, x3
test 1 with model bert-base took  0.186 sec.
test 2 with model bloomz-1b7 took 3.740 sec.
test 1 with model bert-base took  0.208 sec.
test 2 with model bloomz-1b7 took 3.889 sec.
test 1 with model bert-base took  0.187 sec.
test 2 with model bloomz-1b7 took 3.762 sec.

# PR 887, x3
test 1 with model bert-base took  0.019 sec.
test 2 with model bloomz-1b7 took 0.057 sec.
test 1 with model bert-base took  0.019 sec.
test 2 with model bloomz-1b7 took 0.030 sec.
test 1 with model bert-base took  0.019 sec.
test 2 with model bloomz-1b7 took 0.029 sec.
```

So this looks like a very decent speed up, ~10x, although in practice a lot of time is spent on loading the model in the first place, so overall clock time is only 11 vs 7 sec.

> let me know if you prefer to proceed with #887 or with this PR.

My impression is that it would be even better not to initialize the weights at all instead of initializing them on meta device. Having useless inits is just causing confusion. The main advantage with initializing with meta is that the code changes are smaller, so they're less likely to accidentally break something. But from what I can tell, removing the inits should also not break anything.

@younesbelkada @pacman100 Do you have an opinion on that? How should we proceed?

**Comment by poedator** [comment]
> Could you please paste your measurements for reference? 
my results were at the top of the [gist with code](https://gist.github.com/poedator/45331cd9c7837cbfea0ad578b3ce98ed)
let me also copy them here:
```
Switched to branch 'pr887'       # PR #887 by Benjamin Bossan
test 1 with model bert-base took 0.048 sec.
test 2 with model bloomz-1b7 took 0.607 sec.
test 3 with model Llama2-7b took 5.863 sec.

Switched to branch 'fast_init'   # PR #872 by poedator@
test 1 with model bert-base took 0.063 sec.       ~3x
test 2 with model bloomz-1b7 took 0.528 sec.      ~7x
test 3 with model Llama2-7b took 5.618 sec.       ~15x

Switched to branch 'main'         # main version, commit 85013987aa82aa1af3da1236b6902556ce3e483e
test 1 with model bert-base took 0.224 sec.
test 2 with model bloomz-1b7 took 3.842 sec.
test 3 with model Llama2-7b took 76.174 sec.  <---- terribly slow (vs 7 sec for loading whole 7B model)

tested on fast machine with A100 and 16 kernels.
all timing is net of loading time, only get_peft_model()
```

**Review by younesbelkada — APPROVED** [review]
Amazing work @poedator @BenjaminBossan thanks a lot for this! 🚀

**Comment by younesbelkada** [comment]
The current way of proceeding looks great IMO as the code change is quite minimal for a nice gain, also I think `meta` device has been introduced in torch>=1.13 which is the current min requirement for PEFT: https://github.com/huggingface/peft/blob/main/setup.py#L45

**Comment by pacman100** [comment]
Thank you @poedator and @BenjaminBossan for making the LoRA init super fast  🚀. I really like both approaches, favouring Benjamin's as it avoids the inits of weights altogether.

**Comment by regisss** [comment]
Anything else missing for merging this PR?

**Comment by BenjaminBossan** [comment]
Since #887 is merged, I think we can close this one. Progress for other layers/adapters is tracked in #896.
